pkg/model/iam.go

```diff
@@ -72,33 +87,57 @@ func (b *IAMModelBuilder) Build(c *fi.ModelBuilderContext) error {
 
 	// Generate IAM tasks for each shared role
 	for profileARN, igRole := range sharedProfileARNsToIGRole {
+		role := iam.PodOrNodeRole{NodeRole: igRole}
+
 		iamName, err := findCustomAuthNameFromArn(profileARN)
 		if err != nil {
 			return fmt.Errorf("unable to parse instance profile name from arn %q: %v", profileARN, err)
 		}
-		err = b.buildIAMTasks(igRole, iamName, c, true)
+		err = b.buildIAMTasks(role, iamName, c, true)
 		if err != nil {
 			return err
 		}
 	}
 
 	// Generate IAM tasks for each managed role
 	for igRole := range managedRoles {
+		role := iam.PodOrNodeRole{NodeRole: igRole}
 		iamName := b.IAMName(igRole)
-		err := b.buildIAMTasks(igRole, iamName, c, false)
+		err := b.buildIAMTasks(role, iamName, c, false)
 		if err != nil {
 			return err
 		}
 	}
 
+	// Generate IAM tasks for service-account roles
+	if featureflag.UseServiceAccountIAM.Enabled() {
+		serviceAccountRoles := []iam.ServiceAccountRole{iam.ServiceAccountRoleDNSController}
```

Comment:
This doesn't seem a good place for determining which service account roles need to be created. For example, if dns-controller is disabled then we should not build its service account role. That information is more properly in bootstrapchannelbuilder or the respective addon manifest templates.

I'm thinking that BootstrapChannelBuilder could build the IAM tasks, either directly or as a consequence of the manifest template calling a template function with the name of the role as a parameter.

Reply:
Good idea - moved to BootstrapChannelBuilder.

I'm generally trying to avoid magic templates, so I made it explicit.

I would like in future to have the addon able to declare the actual IAM permissions it needs, ideally by having a CRD that represented the IAM policy ... but that would make this change look small!

Reply:
ACK is unfortunately a little underbaked as of yet.